### PR TITLE
Add blog post on extensible broadcast fusion

### PR DESCRIPTION
**Pull request by mbauman**
I'm still working on the package ecosystem examples, but the first few sections are ready for review.

### CODE REVIEW

File: blog/_posts/2018-05-01-extensible-broadcast-fusion.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
++---
+ +layout: post
+ +title:  "Extensible broadcast fusion"
+ +author: <a href="https://github.com/mbauman">Matt Bauman (Julia Computing)</a>
+ +---
+
+Julia version 0.7 brings with it an exciting new feature: the ability to customize broadcast fusion! This recently-merged change is the culmination of a long iterative design process that involved many members of the community. We have converged on a highly extensible interface that should satisfy many use-cases.  In this blog post I'll explain why this is a big deal by reviewing some of the key features and just scratch the surface of all that is possible with this new design. I'm quite certain that our enterprising community will come up with many more clever ways to exploit this new structure in the future.
```
Comment (ararslan) [inline blog/_posts/2018-05-01-extensible-broadcast-fusion.md]: I might suggest using a different word than "exploit," which almost makes this sound like a security risk.

File: blog/_posts/2018-05-01-extensible-broadcast-fusion.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+Julia now uses a first-class datastructure to "lazily" represent a fused broadcast expression before executing it. If you're not a package developer this may not mean all that much to you, but you'll still reap the many rewards. In Base Julia and its standard libraries alone, this means:
```
Comment (ararslan) [inline blog/_posts/2018-05-01-extensible-broadcast-fusion.md]: "datastructure" -> "data structure"

Comment (ararslan) [inline blog/_posts/2018-05-01-extensible-broadcast-fusion.md]: Perhaps worth referencing Steven G. Johnson's post that talks about fused broadcast: https://julialang.org/blog/2017/01/moredots.

File: blog/_posts/2018-05-01-extensible-broadcast-fusion.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+* Ranges can now identify cases where they can compute the result in terms of a new range. For example, the expression `((1:10000) .+ 20) .* 7` doesn't need to allocate a vector for 10,000 elements — it doesn't even need to do 10,000 computations. It can simply operate in terms of the start, stop, and step to compute the result: `147:7:70140`.
```
Comment (ararslan) [inline blog/_posts/2018-05-01-extensible-broadcast-fusion.md]: I'd put `Range`s in backticks like so, similarly for `BitArray`s and perhaps `LinearAlgebra` below.

File: blog/_posts/2018-05-01-extensible-broadcast-fusion.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+Julia computes this result by first constructing an object that represents the "kernel" and then immediately executing it. This little bit of indirection allows a whole host of customizations.
+
+### The representation of a fused broadcast
+
+You can see precisely how a fused broadcast is represented with `Meta.@lower`, but in simpler terms the expression `([1, 2, 3] .+ [10 20 30 40]) ./ 10` is effectively a syntax transformation for:
+
+```julia
+julia> using .Broadcast: materialize, broadcasted
+       bc = broadcasted(/, broadcasted(+, [1, 2, 3], [10 20 30 40]), 10)
+	   materialize(bc)
```
Comment (ararslan) [inline blog/_posts/2018-05-01-extensible-broadcast-fusion.md]: Random tab here

File: blog/_posts/2018-05-01-extensible-broadcast-fusion.md (hunk 1 above)
Comment (Sacha0) [reply blog/_posts/2018-05-01-extensible-broadcast-fusion.md]: Perhaps "leverage"?

**Comment by ViralBShah** [comment]
Suggestion: Would be nice to have a paragraph or two showing what this means for the end user. Seems like you should see lesser memory and better performance on these kinds of expressions, and it may perhaps even be possible to create an example that shows this off. Even better if it can be some small known kernel like a stencil operation or something.

**Comment by ViralBShah** [comment]
I think the `BitArray` example is awesome. In addition could we have a non-bitarray, bit more standard example?

**Comment by mbauman** [comment]
I don't really have any good examples that come to mind with standard `Array`s — they already worked great with broadcast on 0.6.  Anyone have any slightly-more-standard ideas? It's also hard to grab something from the package ecosystem because generally those require large changes to get them up to speed with the new API.

**Comment by mbauman** [comment]
Thanks everyone for the reviews! I pushed a new update last night.

File: blog/_posts/2018-05-01-extensible-broadcast-fusion.md [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+	 101.327  100.0    100.0
+	 100.85   102.138  100.0
+	 100.575  100.129  103.839
+    ```
+
+	Previously, Julia 0.6 would have returned a SparseMatrixCSC for `d ./ 10` and an `Array` for `d .+ LowerTriangular(rand(3,3))`.
+
+* Finally, broadcasting at the global scope is now pre-compilable, and you can use
+  dot-broadcast inside generated functions. This wasn't a huge limitation in the past, but
+  it did cause surprise to folks timing an in-place broadcast like `@time y .*= 2` and seeing thousands of allocations:
```
Comment (Sacha0) [inline blog/_posts/2018-05-01-extensible-broadcast-fusion.md]: Perhaps "did surprise folks timing in-place broadcasts like"?

File: blog/_posts/2018-05-01-extensible-broadcast-fusion.md [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+inlining and argument specialization for near-zero overhead. With this basic framework in
+mind, you can begin to see how the built-in and standard library arrays are able to
+implement all those new features mentioned above:
+
+* When broadcasting into a `BitArray`, it can first introspect the functions and their
+  arguments in the expression tree to see if it can operate at the level of the packed
+  64-bit chunks as `UInt64`s instead of working bit-by-bit. It can even convert boolean-only
+  operators like `!` to their equivalent bitwise operators like `~`.
+
+* Ranges are able to "opt-out" of fusion by defining specialized `Broadcast.broadcasted`
```
Comment (Sacha0) [inline blog/_posts/2018-05-01-extensible-broadcast-fusion.md]: Depending on your answer to the related comment above, perhaps "`Range`s"? :)

File: blog/_posts/2018-05-01-extensible-broadcast-fusion.md [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+  methods that immediately return those re-computed ranges. This means that they do not fuse
+  multiple operations at all, but in exchange they get an $O(1)$ algorithm.
+
+* When LinearAlgebra's structured matrices are asked to allocate the result, their
```
Comment (Sacha0) [inline blog/_posts/2018-05-01-extensible-broadcast-fusion.md]: Perhaps "`LinearAlgebra`"?

File: blog/_posts/2018-05-01-extensible-broadcast-fusion.md [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+	  mean time:        3.764 ms (0.18% GC)
+	  maximum time:     7.744 ms (50.18% GC)
+	  --------------
+	  samples:          1328
+	  evals/sample:     1
+    ```
+
+	That's upwards of a **450x performance gain**.
+
+* Ranges can now compute the result of a broadcast in terms of a new range if it is
```
[the remark this answers is not in the document]
Comment (mbauman) [reply blog/_posts/2018-05-01-extensible-broadcast-fusion.md]: Yes, your edit is exactly correct — I'm trying to slangily refer to `AbstractRange`s with a lower case "range" that happens to be at the start of a sentence.

**Comment by mbauman** [comment]
Alright — I think this is good to go!

**Comment by ViralBShah** [comment]
Can you update the date on the post and merge it?

**Pull request merged**